cdcv2/metadata/sql/tables.go

```diff
@@ -29,11 +32,65 @@ const (
 
 // UpstreamDO mapped from table <upstream>
 type UpstreamDO struct {
-	ID        uint64      `gorm:"column:id;type:bigint(20) unsigned;primaryKey" json:"id"`
-	Endpoints string      `gorm:"column:endpoints;type:text;not null" json:"endpoints"`
-	Config    *Credential `gorm:"column:config;type:text" json:"config"`
-	Version   uint64      `gorm:"column:version;type:bigint(20) unsigned;not null" json:"version"`
-	UpdateAt  time.Time   `gorm:"column:update_at;type:datetime(6);not null;autoUpdateTime" json:"update_at"`
+	ID        uint64               `gorm:"column:id;type:bigint(20) unsigned;primaryKey" json:"id"`
+	Endpoints string               `gorm:"column:endpoints;type:text;not null" json:"endpoints"`
+	Config    *security.Credential `gorm:"column:config;type:text" json:"config"`
+	Version   uint64               `gorm:"column:version;type:bigint(20) unsigned;not null" json:"version"`
+	UpdateAt  time.Time            `gorm:"column:update_at;type:datetime(6);not null;autoUpdateTime" json:"update_at"`
+}
+
+// equal checks whether two UpstreamDO are equal.
+func (u *UpstreamDO) equal(other *UpstreamDO) bool {
+	if u != other && (u == nil || other == nil) {
+		return false
+	}
+
+	if u.ID != other.ID {
+		return false
+	}
+	if u.Endpoints != other.Endpoints {
+		return false
+	}
+
+	if u.Config == other.Config {
+		return true
+	}
+
+	if u.Config == nil && other.Config != nil {
+		return false
+	}
+	if u.Config != nil && other.Config == nil {
+		return false
+	}
+
+	if u.Config.CAPath != other.Config.CAPath ||
+		u.Config.CertPath != other.Config.CertPath ||
+		u.Config.KeyPath != other.Config.KeyPath ||
+		len(u.Config.CertAllowedCN) != len(other.Config.CertAllowedCN) {
+		return false
+	}
+	for i := range u.Config.CertAllowedCN {
```

Comment:
`CertAllowedCN` is not used to connect to upstream.

Reply:
Considering that this field is also persistent, even if it is not currently used, it is necessary to determine whether it is equal here to facilitate future extensions.